pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -191,26 +191,38 @@ func BuildListenerTLSContext(serverTLSSettings *networking.ServerTLSSettings,
 		}
 	}
 
-	// Set TLS parameters if they are non-default
-	var ecdhCurves []string
-	if mesh != nil && mesh.TlsDefaults != nil &&
-		(serverTLSSettings.Mode == networking.ServerTLSSettings_MUTUAL || serverTLSSettings.Mode == networking.ServerTLSSettings_SIMPLE) {
-		ecdhCurves = mesh.TlsDefaults.EcdhCurves
-	}
-	if len(serverTLSSettings.CipherSuites) > 0 || len(ecdhCurves) > 0 ||
-		serverTLSSettings.MinProtocolVersion != networking.ServerTLSSettings_TLS_AUTO ||
-		serverTLSSettings.MaxProtocolVersion != networking.ServerTLSSettings_TLS_AUTO {
-		ctx.CommonTlsContext.TlsParams = &auth.TlsParameters{
-			TlsMinimumProtocolVersion: convertTLSProtocol(serverTLSSettings.MinProtocolVersion),
-			TlsMaximumProtocolVersion: convertTLSProtocol(serverTLSSettings.MaxProtocolVersion),
-			CipherSuites:              serverTLSSettings.CipherSuites,
-			EcdhCurves:                ecdhCurves,
+	// If Mesh TLSDefaults are set, use them.
+	if mesh.GetTlsDefaults() != nil && isSimpleOrMutual(serverTLSSettings.Mode) {
+		if len(mesh.TlsDefaults.EcdhCurves) > 0 {
+			tlsParamsOrNew(ctx.CommonTlsContext).EcdhCurves = mesh.TlsDefaults.EcdhCurves
+		}
+		// Set the minimum TLS version from server settings or mesh config.
+		if serverTLSSettings.MinProtocolVersion != networking.ServerTLSSettings_TLS_AUTO {
+			tlsParamsOrNew(ctx.CommonTlsContext).TlsMinimumProtocolVersion = convertTLSProtocol(serverTLSSettings.MinProtocolVersion)
+		} else if mesh.TlsDefaults.MinProtocolVersion != meshconfig.MeshConfig_TLSConfig_TLS_AUTO {
+			tlsParamsOrNew(ctx.CommonTlsContext).TlsMinimumProtocolVersion = auth.TlsParameters_TlsProtocol(mesh.TlsDefaults.MinProtocolVersion)
 		}
 	}
-
+	if len(serverTLSSettings.CipherSuites) > 0 {
+		tlsParamsOrNew(ctx.CommonTlsContext).CipherSuites = serverTLSSettings.CipherSuites
+	}
+	if serverTLSSettings.MaxProtocolVersion != networking.ServerTLSSettings_TLS_AUTO {
+		tlsParamsOrNew(ctx.CommonTlsContext).TlsMaximumProtocolVersion = convertTLSProtocol(serverTLSSettings.MaxProtocolVersion)
+	}
 	return ctx
 }
 
+func isSimpleOrMutual(mode networking.ServerTLSSettings_TLSmode) bool {
```

Comment:
How about do the negative check, istio_mtls

Reply:
it makes code hard to read at the call site.